File: task_builder/frontend/src/rating.ts
import rules from '../../../backend/rating_rules.json';

const criteria: Record<string, Record<string, number>> = {
  context_and_need: { context: 10, need: 10 }, data: { data: 20 }, expected_result: { expected_result: 15 },
  success_criteria: { success_criteria: 15 }, constraints: { constraints: 10 }, users: { users: 10 },
  business_contact: { contact: 5, interaction_format: 5 },
};
// Mirrors backend/rating.py; this checks readable structure, not factual accuracy.
function readableWords(value: string) {
  return (value.toLowerCase().replace(/ß/g, 'ss').replace(/ς/g, 'σ').match(/\p{L}+/gu) ?? []).filter(word => {
    if (Array.from(word).length < 2 || rules.placeholder_words.includes(word)) return false;
    if (rules.placeholder_words.some(base => word.length % base.length === 0 && base.repeat(word.length / base.length) === word)) return false;
    if ([1, 2, 3].some(size => word.length >= size * 3 && word.length % size === 0 && word.slice(0, size).repeat(word.length / size) === word)) return false;
    return !/^[a-zа-яё]+$/.test(word) || /[aeiouyаеёиоуыэюя]/.test(word) || rules.allowed_acronyms.includes(word);
  });
}
export function fieldIssue(name: string, value: string): string | null {
  const text = value.trim();
  if (!text) return rules.empty_reason;
  const words = readableWords(text);
  if (name === 'contact') {
    if (/^[^@\s]+@(?:[\p{L}\p{N}_-]+\.)+[\p{L}\p{N}_-]{2,}$/u.test(text)) return null;
    const digits = text.match(/\p{Nd}/gu) ?? [];
    if (/^\+?[\p{Nd}()\s.\-]+$/u.test(text) && digits.length >= 7 && digits.length <= 15 && new Set(digits).size > 1) return null;
    if (/^@[a-zA-Z][a-zA-Z0-9_]{4,31}$/.test(text) && words.length) return null;
    try {
      const url = new URL(text);
      if (/^https?:\/\//i.test(text) && ['http:', 'https:'].includes(url.protocol) && url.hostname && !text.split('/')[2]?.includes('@') && !/[\s\x00-\x1f\\]/.test(text)) return null;
    } catch { /* A contact can still be a person's name. */ }
    return new Set(words).size >= rules.min_distinct_words && text.split(/\s+/).length >= 2 && /^\p{L}+(?:[\s'’-]\p{L}+)+$/u.test(text) ? null : rules.contact_reason;
  }
  if (!words.length) return rules.placeholder_reason;
  return new Set(words).size < rules.min_distinct_words || words.reduce((total, word) => total + Array.from(word).length, 0) < rules.min_letters ? rules.detail_reason : null;
}
export function evaluatePreview(task: Record<string, unknown>) {
  const result = { score: 0, level: 'draft', score_breakdown: {} as Record<string, { earned: number; max: number }>, missing_fields: [] as string[], field_issues: {} as Record<string, string> };
  for (const [criterion, parts] of Object.entries(criteria)) {
    let earned = 0; let max = 0;
    for (const [field, weight] of Object.entries(parts)) {
      max += weight;
      const issue = fieldIssue(field, String(task[field] ?? ''));
      if (issue) { result.missing_fields.push(field); result.field_issues[field] = issue; }
      else earned += weight;
    }
    result.score += earned; result.score_breakdown[criterion] = { earned, max };
  }
  result.level = result.score < 40 ? 'draft' : result.score < 70 ? 'working' : result.score < 90 ? 'ready' : 'priority';
  return result;
}

File: backend/rating.py
"""Rating of business-confirmed task fields."""

from datetime import datetime, timezone
import json
from pathlib import Path
import re
from typing import Mapping
import sqlite3
from urllib.parse import urlparse


RULES = json.loads(Path(__file__).with_name("rating_rules.json").read_text(encoding="utf-8"))
PLACEHOLDERS = set(RULES["placeholder_words"])
ACRONYMS = set(RULES["allowed_acronyms"])


def readable_words(value: str) -> list[str]:
    """A transparent shape check, not a semantic or factual quality assessment."""
    result = []
    for word in re.findall(r"[^\W\d_]+", value.casefold()):
        if len(word) < 2 or word in PLACEHOLDERS:
            continue
        if any(len(word) % len(base) == 0 and word == base * (len(word) // len(base)) for base in PLACEHOLDERS):
            continue
        if any(len(word) >= size * 3 and len(word) % size == 0 and word == word[:size] * (len(word) // size) for size in (1, 2, 3)):
            continue
        if re.fullmatch(r"[a-zа-яё]+", word) and not re.search(r"[aeiouyаеёиоуыэюя]", word) and word not in ACRONYMS:
            continue
        result.append(word)
    return result


def field_issue(name: str, value: object) -> str | None:
    text = str(value or "").strip()
    if not text:
        return RULES["empty_reason"]
    words = readable_words(text)
    if name == "contact":
        if re.fullmatch(r"[^@\s]+@(?:[\w-]+\.)+[\w-]{2,}", text):
            return None
        digits = re.sub(r"\D", "", text)
        if re.fullmatch(r"\+?[\d()\s.\-]+", text) and 7 <= len(digits) <= 15 and len(set(digits)) > 1:
            return None
        if re.fullmatch(r"@[a-zA-Z][a-zA-Z0-9_]{4,31}", text) and words:
            return None
        try:
            url = urlparse(text)
            if (url.scheme in ("http", "https") and url.hostname and url.username is None
                    and "\\" not in text and not any(character.isspace() or ord(character) < 32 for character in text)):
                url.port
                return None
        except ValueError:
            pass
        if len(text.split()) >= 2 and re.fullmatch(r"[^\W\d_]+(?:[\s'’-][^\W\d_]+)+", text) and len(set(words)) >= RULES["min_distinct_words"]:
            return None
        return RULES["contact_reason"]
    if not words:
        return RULES["placeholder_reason"]
    if len(set(words)) < RULES["min_distinct_words"] or sum(map(len, words)) < RULES["min_letters"]:
        return RULES["detail_reason"]
    return None


CRITERIA = {
    "context_and_need": {"context": 10, "need": 10},
    "data": {"data": 20},
    "expected_result": {"expected_result": 15},
    "success_criteria": {"success_criteria": 15},
    "constraints": {"constraints": 10},
    "users": {"users": 10},
    "business_contact": {"contact": 5, "interaction_format": 5},
}
EDITABLE_FIELDS = (
    "title", "topic", "context", "need", "users", "data", "constraints",
    "expected_result", "success_criteria", "contact", "interaction_format",
)


def level_for(score: int) -> str:
    if score < 40:
        return "draft"
    if score < 70:
        return "working"
    if score < 90:
        return "ready"
    return "priority"


def evaluate(task: Mapping, confirmed: bool = True) -> dict:
    breakdown = {}
    missing = []
    issues = {}
    score = 0
    for criterion, fields in CRITERIA.items():
        earned = 0
        for name, weight in fields.items():
            issue = field_issue(name, task[name])
            if issue is None:
                if confirmed:
                    earned += weight
            else:
                missing.append(name)
                issues[name] = issue
        maximum = sum(fields.values())
        breakdown[criterion] = {"earned": earned, "max": maximum}
        score += earned
    return {
        "score": score,
        "level": level_for(score),
        "score_breakdown": breakdown,
        "missing_fields": missing,
        "field_issues": issues,
    }


def recalculate_task(connection: sqlite3.Connection, task_id: int) -> dict:
    """Persist the score after a confirmed edit; call inside the caller's transaction."""
    row = connection.execute("SELECT * FROM tasks WHERE id = ?", (task_id,)).fetchone()
    if row is None:
        raise ValueError(f"Task {task_id} does not exist")
    result = evaluate(row, confirmed=bool(row["confirmed_at"]))
    connection.execute("UPDATE tasks SET score = ? WHERE id = ?", (result["score"], task_id))
    return result


def confirm_task(connection: sqlite3.Connection, task_id: int, updates: Mapping) -> dict:
    """Save business-approved field changes and recalculate in the same transaction."""
    invalid = set(updates) - set(EDITABLE_FIELDS)
    if invalid:
        raise ValueError(f"Fields cannot be confirmed: {', '.join(sorted(invalid))}")
    if connection.execute("SELECT 1 FROM tasks WHERE id = ?", (task_id,)).fetchone() is None:
        raise ValueError(f"Task {task_id} does not exist")
    if updates:
        columns = ", ".join(f"{name} = ?" for name in updates)
        connection.execute(
            f"UPDATE tasks SET {columns} WHERE id = ?",
            (*updates.values(), task_id),
        )
    connection.execute(
        "UPDATE tasks SET confirmed_at = ? WHERE id = ?",
        (datetime.now(timezone.utc).isoformat(), task_id),
    )
    return recalculate_task(connection, task_id)


def reconcile_ratings(connection: sqlite3.Connection) -> int:
    """Refresh persisted scores with current rules without changing users' text."""
    changed = 0
    for row in connection.execute("SELECT * FROM tasks").fetchall():
        score = evaluate(row, confirmed=bool(row["confirmed_at"]))["score"]
        if score != row["score"]:
            connection.execute("UPDATE tasks SET score = ? WHERE id = ?", (score, row["id"]))
            changed += 1
    return changed
